File: python-client/src/par_rt_db/value_expr.py
"""The closed ``ValueExpr`` grammar (ENH-020 / ENH-028) — the typed,
injection-safe expression language shared by migrate's ``evalExpr`` backfill
and computed fields.

Mirrors ``server/src/value_expr.rs`` (the canonical home of the wire types —
``migrate.rs`` re-exports them there, exactly as :mod:`par_rt_db.migration`
re-exports them here). One wire shape, two executions: the server's SQL
compiler for the one-shot migrate UPDATE, and the in-memory interpreter
(:mod:`par_rt_db.in_memory.value_expr`) for per-write stamping in the engine.

Wire shapes (load-bearing — match the server exactly):

* ``Cast`` is the closed set of sound coercions shared by
  ``changeType`` and ``ValueExpr.cast`` (``toString`` / ``toNumber`` /
  ``toInt64`` / ``toBoolean``; camelCase on the wire).
* ``ValueExpr`` is tagged by ``op`` (camelCase variants: ``field`` /
  ``literal`` / ``concat`` / ``add`` / ``sub`` / ``mul`` / ``div`` /
  ``coalesce`` / ``lower`` / ``upper`` / ``trim`` / ``cast`` / ``now`` /
  ``case``), ``deny_unknown_fields`` via ``extra="forbid"`` — the same serde
  conventions as ``wire.FilterExpr``. The grammar is closed (no subquery /
  function-by-name / raw-SQL node) so the SEC-107 injection concern cannot
  arise from a ``ValueExpr`` payload.
* ``CaseWhen`` is one branch of ``ValueExpr.case``: ``{when: FilterExpr,
  then: ValueExpr}`` (camelCase, ``extra="forbid"``).

The ``ve`` namespace is the DSL constructor surface (``ve.field("name")``,
``ve.concat(...)`` ...), mirroring ``schema.t`` for field types and the rust
client's ``ValueExpr::field``-style associated constructors.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import Field

from .wire import FilterExpr, _Camel


class Cast(StrEnum):
    """Closed set of sound coercions for ``Directive.changeType``.

    Mirrors ``server/src/value_expr.rs::Cast`` (camelCase on the wire).
    ``StrEnum`` so pydantic v2 serializes the string value directly and
    ``Cast.TO_STRING == "toString"`` for ergonomic builder calls. Shared by
    ``changeType`` and ``ValueExpr.cast`` (the four scalar casts sound to
    backfill).
    """

    TO_STRING = "toString"
    TO_NUMBER = "toNumber"
    TO_INT64 = "toInt64"
    TO_BOOLEAN = "toBoolean"


# --- ValueExpr (discriminator "op", camelCase) ---
#
# Closed, typed expression grammar for ``Directive.EvalExpr.expr`` (ENH-020
# Stage 1, closing SEC-107) and ``TableDef.computed`` entries (ENH-028).
# Mirrors ``server/src/value_expr.rs::ValueExpr`` byte-for-byte:
# ``tag = "op"``, camelCase, ``deny_unknown_fields`` (via ``extra="forbid"``
# on ``_Camel``) — the same serde conventions as ``wire.FilterExpr``. Every
# ``literal`` compiles to a bound ``$n`` placeholder (as jsonb); every
# ``field`` resolves through the table's ``TableDef``.
#
# Self-referential union pattern mirrors ``wire.FilterExpr``: ``from __future__
# import annotations`` makes every annotation a string, the union references
# the variant classes by name, and ``model_rebuild()`` at module foot resolves
# the forward refs into a fully-built schema.


class _ValueField(_Camel):
    op: Literal["field"] = "field"
    field: str


class _ValueLiteral(_Camel):
    op: Literal["literal"] = "literal"
    value: Any


class _ValueConcat(_Camel):
    op: Literal["concat"] = "concat"
    parts: list[ValueExpr]


class _ValueAdd(_Camel):
    op: Literal["add"] = "add"
    left: ValueExpr
    right: ValueExpr


class _ValueSub(_Camel):
    op: Literal["sub"] = "sub"
    left: ValueExpr
    right: ValueExpr


class _ValueMul(_Camel):
    op: Literal["mul"] = "mul"
    left: ValueExpr
    right: ValueExpr


class _ValueDiv(_Camel):
    op: Literal["div"] = "div"
    left: ValueExpr
    right: ValueExpr


class _ValueCoalesce(_Camel):
    op: Literal["coalesce"] = "coalesce"
    parts: list[ValueExpr]


class _ValueLower(_Camel):
    op: Literal["lower"] = "lower"
    value: ValueExpr


class _ValueUpper(_Camel):
    op: Literal["upper"] = "upper"
    value: ValueExpr


class _ValueTrim(_Camel):
    op: Literal["trim"] = "trim"
    value: ValueExpr


class _ValueCast(_Camel):
    op: Literal["cast"] = "cast"
    value: ValueExpr
    to: Cast


class _ValueNow(_Camel):
    op: Literal["now"] = "now"


class _ValueCase(_Camel):
    op: Literal["case"] = "case"
    whens: list[CaseWhen]
    otherwise: ValueExpr


class CaseWhen(_Camel):
    """One branch of ``ValueExpr.case``. Wire shape ``{when, then}``.

    Mirrors ``server/src/value_expr.rs::CaseWhen`` (camelCase,
    ``deny_unknown_fields`` via ``extra="forbid"`` on ``_Camel``). ``when`` is
    the read-path ``FilterExpr`` (field references schema-validated, values
    bound); ``then`` is the typed expression result on a match.
    """

    when: FilterExpr
    then: ValueExpr


#: Discriminated union of all 14 ``ValueExpr`` ops. The ``op`` literal drives
#: dispatch; ``deny_unknown_fields`` is per-variant via ``extra="forbid"`` on
#: ``_Camel`` — the same shape contract as ``wire.FilterExpr``.
ValueExpr = Annotated[
    (
        _ValueField
        | _ValueLiteral
        | _ValueConcat
        | _ValueAdd
        | _ValueSub
        | _ValueMul
        | _ValueDiv
        | _ValueCoalesce
        | _ValueLower
        | _ValueUpper
        | _ValueTrim
        | _ValueCast
        | _ValueNow
        | _ValueCase
    ),
    Field(discriminator="op"),
]

# Resolve deferred annotations (``from __future__ import annotations`` makes
# every annotation a string; ``model_rebuild`` evaluates them so the
# discriminated-union schema is fully built before first use). ``ValueExpr``
# is mutually recursive with ``CaseWhen`` (``_ValueCase`` -> ``CaseWhen`` ->
# ``ValueExpr``); rebuilding ``CaseWhen`` first ensures both arms of the cycle
# resolve.
CaseWhen.model_rebuild()


# --- ``ve`` expression constructors ---
#
# Constructors return the wire-identical dict (the simplest form, mirroring
# ``schema.t``); ``SchemaDef.model_validate`` routes each through the
# discriminated ``ValueExpr`` union at ``build()`` time. ``case`` accepts
# ``CaseWhen`` models or plain ``{"when": ..., "then": ...}`` dicts.


class _VeNamespace:
    """Expression constructors (``ve.field(name)`` ... ``ve.case(...)``)."""

    @staticmethod
    def field(name: str) -> dict[str, Any]:
        """Read declared field ``name`` as text (the ``doc->>'field'`` read)."""
        return {"op": "field", "field": name}

    @staticmethod
    def literal(value: Any) -> dict[str, Any]:
        """A constant JSON value (string/number/bool/object/array/null)."""
        return {"op": "literal", "value": value}

    @staticmethod
    def concat(*parts: Any) -> dict[str, Any]:
        """Concatenate the text form of ``parts``, skipping nulls (Postgres
        ``concat()`` — wrap operands in :meth:`coalesce` for explicit control)."""
        return {"op": "concat", "parts": list(parts)}

    @staticmethod
    def add(left: Any, right: Any) -> dict[str, Any]:
        """``left + right`` as IEEE doubles; a null operand nulls the result."""
        return {"op": "add", "left": left, "right": right}

    @staticmethod
    def sub(left: Any, right: Any) -> dict[str, Any]:
        """``left - right`` as IEEE doubles; a null operand nulls the result."""
        return {"op": "sub", "left": left, "right": right}

    @staticmethod
    def mul(left: Any, right: Any) -> dict[str, Any]:
        """``left * right`` as IEEE doubles; a null operand nulls the result."""
        return {"op": "mul", "left": left, "right": right}

    @staticmethod
    def div(left: Any, right: Any) -> dict[str, Any]:
        """``left / right`` as IEEE doubles; a null operand nulls the result
        (null propagation precedes the division-by-zero error)."""
        return {"op": "div", "left": left, "right": right}

    @staticmethod
    def coalesce(*parts: Any) -> dict[str, Any]:
        """First non-null part, else null."""
        return {"op": "coalesce", "parts": list(parts)}

    @staticmethod
    def lower(value: Any) -> dict[str, Any]:
        """Lowercase the text form of ``value`` (null passes through)."""
        return {"op": "lower", "value": value}

    @staticmethod
    def upper(value: Any) -> dict[str, Any]:
        """Uppercase the text form of ``value`` (null passes through)."""
        return {"op": "upper", "value": value}

    @staticmethod
    def trim(value: Any) -> dict[str, Any]:
        """Strip leading/trailing SPACES ONLY from the text form of ``value``
        (Postgres ``btrim`` default — not full Unicode whitespace)."""
        return {"op": "trim", "value": value}

    @staticmethod
    def cast(value: Any, to: Cast) -> dict[str, Any]:
        """Apply the closed scalar coercion ``to`` (one of :class:`Cast`)."""
        return {"op": "cast", "value": value, "to": to}

    @staticmethod
    def now() -> dict[str, Any]:
        """Current epoch-milliseconds as a JSON number (the same value the
        engine stamps for ``updatedAtField``)."""
        return {"op": "now"}

    @staticmethod
    def case(whens: list[Any], otherwise: Any) -> dict[str, Any]:
        """First matching ``when``'s ``then``, else ``otherwise``. Each
        ``whens`` entry is a :class:`CaseWhen` (or the plain
        ``{"when": ..., "then": ...}`` dict)."""
        return {"op": "case", "whens": list(whens), "otherwise": otherwise}


ve = _VeNamespace()

File: go-client/wire/valueexpr.go
// go-client/wire/valueexpr.go
package wire

// Mirrors core/src/wire.rs::ValueExpr — the closed typed expression grammar
// for computed fields and migrate backfill. Tag key "op", camelCase tags,
// camelCase Cast values. Strict decode; interface-typed fields decode
// through per-variant UnmarshalJSON methods.

import (
	"encoding/json"
	"fmt"
)

// Mirrors core/src/wire.rs::Cast — closed scalar coercions, camelCase.
type Cast string

const (
	CastToString  Cast = "toString"
	CastToNumber  Cast = "toNumber"
	CastToInt64   Cast = "toInt64"
	CastToBoolean Cast = "toBoolean"
)

// ValueExpr is the closed, typed expression grammar. Variants:
// ValueField/Literal/Concat/Add/Sub/Mul/Div/Coalesce/Lower/Upper/Trim/Cast/
// Now/Case. There is deliberately no subquery, no function-call-by-name,
// and no raw-SQL node — the grammar is closed.
type ValueExpr interface {
	isValueExpr()
}

// Mirrors core/src/wire.rs::ValueExpr::Field
type ValueField struct {
	Field string `json:"field"`
}

func (ValueField) isValueExpr() {}

func (v ValueField) MarshalJSON() ([]byte, error) {
	type alias ValueField
	return MarshalTagged("op", "field", alias(v))
}

func (v *ValueField) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Field string `json:"field"`
	}](b)
	if err != nil {
		return err
	}
	v.Field = r.Field
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Literal
type ValueLiteral struct {
	Value JSONValue `json:"value"`
}

func (ValueLiteral) isValueExpr() {}

func (v ValueLiteral) MarshalJSON() ([]byte, error) {
	type alias ValueLiteral
	return MarshalTagged("op", "literal", alias(v))
}

func (v *ValueLiteral) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Value json.RawMessage `json:"value"`
	}](b)
	if err != nil {
		return err
	}
	val, err := UnmarshalJSON(r.Value)
	if err != nil {
		return err
	}
	v.Value = val
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Concat
type ValueConcat struct {
	Parts []ValueExpr `json:"parts"`
}

func (ValueConcat) isValueExpr() {}

func (v ValueConcat) MarshalJSON() ([]byte, error) {
	type alias ValueConcat
	return MarshalTagged("op", "concat", alias(v))
}

func (v *ValueConcat) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Parts []json.RawMessage `json:"parts"`
	}](b)
	if err != nil {
		return err
	}
	parts, err := decodeValueExprs(r.Parts)
	if err != nil {
		return err
	}
	v.Parts = parts
	return nil
}

// decodeValueExprs converts raw JSON objects into ValueExpr variants.
func decodeValueExprs(raws []json.RawMessage) ([]ValueExpr, error) {
	out := make([]ValueExpr, len(raws))
	for i := range raws {
		e, err := UnmarshalValueExpr(raws[i])
		if err != nil {
			return nil, err
		}
		out[i] = e
	}
	return out, nil
}

// Mirrors core/src/wire.rs::ValueExpr::Add
type ValueAdd struct {
	Left  ValueExpr `json:"left"`
	Right ValueExpr `json:"right"`
}

func (ValueAdd) isValueExpr() {}

func (v ValueAdd) MarshalJSON() ([]byte, error) {
	type alias ValueAdd
	return MarshalTagged("op", "add", alias(v))
}

func (v *ValueAdd) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Left  json.RawMessage `json:"left"`
		Right json.RawMessage `json:"right"`
	}](b)
	if err != nil {
		return err
	}
	left, err := UnmarshalValueExpr(r.Left)
	if err != nil {
		return err
	}
	right, err := UnmarshalValueExpr(r.Right)
	if err != nil {
		return err
	}
	v.Left, v.Right = left, right
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Sub
type ValueSub struct {
	Left  ValueExpr `json:"left"`
	Right ValueExpr `json:"right"`
}

func (ValueSub) isValueExpr() {}

func (v ValueSub) MarshalJSON() ([]byte, error) {
	type alias ValueSub
	return MarshalTagged("op", "sub", alias(v))
}

func (v *ValueSub) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Left  json.RawMessage `json:"left"`
		Right json.RawMessage `json:"right"`
	}](b)
	if err != nil {
		return err
	}
	left, err := UnmarshalValueExpr(r.Left)
	if err != nil {
		return err
	}
	right, err := UnmarshalValueExpr(r.Right)
	if err != nil {
		return err
	}
	v.Left, v.Right = left, right
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Mul
type ValueMul struct {
	Left  ValueExpr `json:"left"`
	Right ValueExpr `json:"right"`
}

func (ValueMul) isValueExpr() {}

func (v ValueMul) MarshalJSON() ([]byte, error) {
	type alias ValueMul
	return MarshalTagged("op", "mul", alias(v))
}

func (v *ValueMul) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Left  json.RawMessage `json:"left"`
		Right json.RawMessage `json:"right"`
	}](b)
	if err != nil {
		return err
	}
	left, err := UnmarshalValueExpr(r.Left)
	if err != nil {
		return err
	}
	right, err := UnmarshalValueExpr(r.Right)
	if err != nil {
		return err
	}
	v.Left, v.Right = left, right
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Div
type ValueDiv struct {
	Left  ValueExpr `json:"left"`
	Right ValueExpr `json:"right"`
}

func (ValueDiv) isValueExpr() {}

func (v ValueDiv) MarshalJSON() ([]byte, error) {
	type alias ValueDiv
	return MarshalTagged("op", "div", alias(v))
}

func (v *ValueDiv) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Left  json.RawMessage `json:"left"`
		Right json.RawMessage `json:"right"`
	}](b)
	if err != nil {
		return err
	}
	left, err := UnmarshalValueExpr(r.Left)
	if err != nil {
		return err
	}
	right, err := UnmarshalValueExpr(r.Right)
	if err != nil {
		return err
	}
	v.Left, v.Right = left, right
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Coalesce
type ValueCoalesce struct {
	Parts []ValueExpr `json:"parts"`
}

func (ValueCoalesce) isValueExpr() {}

func (v ValueCoalesce) MarshalJSON() ([]byte, error) {
	type alias ValueCoalesce
	return MarshalTagged("op", "coalesce", alias(v))
}

func (v *ValueCoalesce) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Parts []json.RawMessage `json:"parts"`
	}](b)
	if err != nil {
		return err
	}
	parts, err := decodeValueExprs(r.Parts)
	if err != nil {
		return err
	}
	v.Parts = parts
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Lower
type ValueLower struct {
	Value ValueExpr `json:"value"`
}

func (ValueLower) isValueExpr() {}

func (v ValueLower) MarshalJSON() ([]byte, error) {
	type alias ValueLower
	return MarshalTagged("op", "lower", alias(v))
}

func (v *ValueLower) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Value json.RawMessage `json:"value"`
	}](b)
	if err != nil {
		return err
	}
	e, err := UnmarshalValueExpr(r.Value)
	if err != nil {
		return err
	}
	v.Value = e
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Upper
type ValueUpper struct {
	Value ValueExpr `json:"value"`
}

func (ValueUpper) isValueExpr() {}

func (v ValueUpper) MarshalJSON() ([]byte, error) {
	type alias ValueUpper
	return MarshalTagged("op", "upper", alias(v))
}

func (v *ValueUpper) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Value json.RawMessage `json:"value"`
	}](b)
	if err != nil {
		return err
	}
	e, err := UnmarshalValueExpr(r.Value)
	if err != nil {
		return err
	}
	v.Value = e
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Trim
type ValueTrim struct {
	Value ValueExpr `json:"value"`
}

func (ValueTrim) isValueExpr() {}

func (v ValueTrim) MarshalJSON() ([]byte, error) {
	type alias ValueTrim
	return MarshalTagged("op", "trim", alias(v))
}

func (v *ValueTrim) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Value json.RawMessage `json:"value"`
	}](b)
	if err != nil {
		return err
	}
	e, err := UnmarshalValueExpr(r.Value)
	if err != nil {
		return err
	}
	v.Value = e
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Cast
type ValueCast struct {
	Value ValueExpr `json:"value"`
	To    Cast      `json:"to"`
}

func (ValueCast) isValueExpr() {}

func (v ValueCast) MarshalJSON() ([]byte, error) {
	type alias ValueCast
	return MarshalTagged("op", "cast", alias(v))
}

func (v *ValueCast) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Value json.RawMessage `json:"value"`
		To    Cast            `json:"to"`
	}](b)
	if err != nil {
		return err
	}
	e, err := UnmarshalValueExpr(r.Value)
	if err != nil {
		return err
	}
	v.Value, v.To = e, r.To
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Now — unit variant, wire form
// {"op":"now"}.
type ValueNow struct{}

func (ValueNow) isValueExpr() {}

func (v ValueNow) MarshalJSON() ([]byte, error) { return MarshalTagged("op", "now", struct{}{}) }

func (v *ValueNow) UnmarshalJSON(b []byte) error {
	_, err := StrictUnmarshal[struct{}](b)
	return err
}

// Mirrors core/src/wire.rs::CaseWhen — one branch of Case. Wire shape
// {when, then}.
type CaseWhen struct {
	When FilterExpr `json:"when"`
	Then ValueExpr  `json:"then"`
}

func (c *CaseWhen) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		When json.RawMessage `json:"when"`
		Then json.RawMessage `json:"then"`
	}](b)
	if err != nil {
		return err
	}
	w, err := UnmarshalFilterExpr(r.When)
	if err != nil {
		return err
	}
	t, err := UnmarshalValueExpr(r.Then)
	if err != nil {
		return err
	}
	c.When, c.Then = w, t
	return nil
}

// Mirrors core/src/wire.rs::ValueExpr::Case
type ValueCase struct {
	Whens     []CaseWhen `json:"whens"`
	Otherwise ValueExpr  `json:"otherwise"`
}

func (ValueCase) isValueExpr() {}

func (v ValueCase) MarshalJSON() ([]byte, error) {
	type alias ValueCase
	return MarshalTagged("op", "case", alias(v))
}

func (v *ValueCase) UnmarshalJSON(b []byte) error {
	r, err := StrictUnmarshal[struct {
		Whens     []json.RawMessage `json:"whens"`
		Otherwise json.RawMessage   `json:"otherwise"`
	}](b)
	if err != nil {
		return err
	}
	whens := make([]CaseWhen, len(r.Whens))
	for i := range r.Whens {
		if err := json.Unmarshal(r.Whens[i], &whens[i]); err != nil {
			return err
		}
	}
	otherwise, err := UnmarshalValueExpr(r.Otherwise)
	if err != nil {
		return err
	}
	v.Whens, v.Otherwise = whens, otherwise
	return nil
}

// UnmarshalValueExpr routes a tagged expr object to its variant. Unknown
// op is an error, mirroring serde's closed enum.
func UnmarshalValueExpr(data []byte) (ValueExpr, error) {
	tag, err := PeekTag(data, "op")
	if err != nil {
		return nil, err
	}
	switch tag {
	case "field":
		return DecodeTagged[ValueField](data, "op")
	case "literal":
		return DecodeTagged[ValueLiteral](data, "op")
	case "concat":
		return DecodeTagged[ValueConcat](data, "op")
	case "add":
		return DecodeTagged[ValueAdd](data, "op")
	case "sub":
		return DecodeTagged[ValueSub](data, "op")
	case "mul":
		return DecodeTagged[ValueMul](data, "op")
	case "div":
		return DecodeTagged[ValueDiv](data, "op")
	case "coalesce":
		return DecodeTagged[ValueCoalesce](data, "op")
	case "lower":
		return DecodeTagged[ValueLower](data, "op")
	case "upper":
		return DecodeTagged[ValueUpper](data, "op")
	case "trim":
		return DecodeTagged[ValueTrim](data, "op")
	case "cast":
		return DecodeTagged[ValueCast](data, "op")
	case "now":
		return DecodeTagged[ValueNow](data, "op")
	case "case":
		return DecodeTagged[ValueCase](data, "op")
	default:
		return nil, fmt.Errorf("wire: unknown value op %q", tag)
	}
}
